cmd/kubectl-directpv/migrate.go

```diff
@@ -0,0 +1,382 @@
+					if !send(ListDriveResult{Drive: item}) {
+						return
+					}
+				}
+			}
+
+			if result.Continue == "" {
+				return
+			}
+
+			options.Continue = result.Continue
+		}
+	}()
+
+	return resultCh
+}
+
+// ListVolumes is for listing volumes
+func ListVolumes(ctx context.Context) <-chan ListVolumeResult {
+	resultCh := make(chan ListVolumeResult)
+	go func() {
+		defer close(resultCh)
+
+		send := func(result ListVolumeResult) bool {
+			select {
+			case <-ctx.Done():
+				return false
+			case resultCh <- result:
+				return true
+			}
+		}
+
+		options := metav1.ListOptions{Limit: 1000}
+		for {
+			result, err := directCSIVolumeClient.List(ctx, options)
+			if err != nil {
+				send(ListVolumeResult{Err: err})
+				return
+			}
+
+			for _, item := range result.Items {
+				if !send(ListVolumeResult{Volume: item}) {
+					return
+				}
+			}
+
+			if result.Continue == "" {
+				return
+			}
+
+			options.Continue = result.Continue
+		}
+	}()
+
+	return resultCh
+}
+
+func validateMigrateCmd() error {
+	return nil
+}
+
+func migrateVolumeCRD(oldVolumeList []directcsi.DirectCSIVolume) (newVolumeList []directpv.DirectPVVolume, theerror error) {
+	newVolumes := []directpv.DirectPVVolume{}
+	for _, OldVolume := range oldVolumeList {
+		// FIXME: use types.NewDrive() instead of creating structure.
+		NewVolume := directpv.DirectPVVolume{}
+
+		// TypeMeta fields
+		NewVolume.TypeMeta.Kind = VolumeTypeMetaKind
+		NewVolume.TypeMeta.APIVersion = APIVersion
+
+		// ObjectMeta fields
+		NewVolume.ObjectMeta = OldVolume.ObjectMeta
+
+		// Status fields
+		NewVolume.Status.DataPath = OldVolume.Status.HostPath
+		NewVolume.Status.StagingTargetPath = OldVolume.Status.StagingPath
+		NewVolume.Status.TargetPath = OldVolume.Status.ContainerPath
+		NewVolume.Status.FSUUID = driveFsuuid[OldVolume.Status.Drive]
+		NewVolume.Status.TotalCapacity = OldVolume.Status.TotalCapacity
+		NewVolume.Status.AvailableCapacity = OldVolume.Status.AvailableCapacity
+		NewVolume.Status.UsedCapacity = OldVolume.Status.UsedCapacity
+
+		// Append the new drive to the list
+		newVolumes = append(newVolumes, NewVolume)
+	}
+	return newVolumes, nil
+}
+
+func migrateDriveCRD(oldDriveList []directcsi.DirectCSIDrive) (newDriveList []directpv.DirectPVDrive, therror error) {
+	// Set of unique FSUUIDs:
+	setOfUniqueFSUUIDs := make(map[string]struct{})
+
+	newDrives := []directpv.DirectPVDrive{}
+	for _, OldDrive := range oldDriveList {
+
+		// we can filter out the drives other than Ready and InUse
+		if OldDrive.Status.DriveStatus != directcsi.DriveStatusReady && OldDrive.Status.DriveStatus != directcsi.DriveStatusInUse {
+			continue // skip this iteration to filter out this state.
+		}
+
+		// FIXME: use types.NewVolume() instead of creating structure.
+		NewDrive := directpv.DirectPVDrive{}
+
+		// TypeMeta fields
+		NewDrive.TypeMeta.Kind = DriveTypeMetaKind
+		NewDrive.TypeMeta.APIVersion = APIVersion
+
+		// ObjectMeta fields
+		NewDrive.ObjectMeta = OldDrive.ObjectMeta
```

Comment:
`ObjectMeta.Finalizers` requires a special treatment. The older finalizers will have `direct.csi.min.io/` as its prefix. We need to change it to `directpv.min.io/`

The same for volumes.